### PR TITLE
ICU-22116 Update CI job for ICU4J to use Java 8 instead of Java 7

### PR DESCRIPTION
**Pull request by echeran**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22116
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by echeran** [comment]
Note: This is a straightforward change from Java 7 to Java 8, and things seem to pass.

I can also update the CI job step that automatically installs Java for us.  Newer versions of that Java setup job step want us to also specify the [name of the distribution of OpenJDK](https://github.com/actions/setup-java#supported-distributions) that we use (not just the version number).  If you want, I can do the update of the Java setup step, including that preference, here too. I don't think people had an opinion on the OpenJDK distribution when I asked previously, and given the advice of Adopt OpenJDK users to use Eclipse Temurin OpenJDK, I would just use 

cc @mihnita

**Comment by markusicu** [comment]
> I can also update the CI job step that automatically installs Java for us. Newer versions of that Java setup job step want us to also specify the [name of the distribution of OpenJDK](https://github.com/actions/setup-java#supported-distributions) that we use (not just the version number). If you want, I can do the update of the Java setup step, including that preference, here too. I don't think people had an opinion on the OpenJDK distribution when I asked previously, and given the advice of Adopt OpenJDK users to use Eclipse Temurin OpenJDK, I would just use

Ok. I didn't actually know that there are multiple "distributions" of OpenJDK... given the explanation, and the GH actions docs, temurin sounds fine.

File: .github/workflows/icu_ci.yml [hunk 1]
```diff
@@ -43,27 +43,6 @@ jobs:
       - uses: actions/setup-java@v1
         with:
           java-version: '8'
```
Comment (markusicu) [inline .github/workflows/icu_ci.yml:45]: I am surprised that this line already said '8' ...

**Review by markusicu — APPROVED** [review]
lgtm tnx!

Happy to re-approve for more changes, such as the distribution.

We should probably think about some piece of code that would benefit from Java 8, to demonstrate that it actually works. Maybe something in the tests first?

File: .github/workflows/icu_ci.yml (hunk 1 above)
Comment (gnrunge) [reply .github/workflows/icu_ci.yml:45]: Further down in that file we test ICU4J with java-version: '11' and java-version: '16', respectively.

**Review by gnrunge — APPROVED** [review]

**Comment by echeran** [comment]
> We should probably think about some piece of code that would benefit from Java 8, to demonstrate that it actually works. Maybe something in the tests first?

Done. I updated a test using the Java 8 Streams/Functions constructs.  

In general, there are tradeoffs of this slightly more functional style, of course.  The benefits are more readable code, less code, less code with manual iteration, so less surface area for mistakes (ex: off-by-one errors).  The costs would be performance -- you're (probably?) creating and GC'ing a lot of objects.  Even though a lot of the Java GCs coming out since Java 8 are tuned to perform better for lots of small objects, to help with Java 8 and the many new FP languages built on the JVM, it makes sense to start making such Java 8 changes by putting them in tests.  I don't know what the perf impacts would be for such changes in hot code paths in the main code (ex: would function/closure literals get translated into local variables at compile time, and would that mitigate the risks much?)

**Review by markusicu — APPROVED** [review]
lgtm / nice! / tnx!

**Pull request merged**